lib/vtls/openssl.c

```diff
@@ -5157,23 +5157,39 @@ static CURLcode ossl_get_tls_server_end_point(struct Curl_easy *data,
 
   /* https://datatracker.ietf.org/doc/html/rfc5929#section-4.1 */
   if(algo_nid == NID_md5 || algo_nid == NID_sha1) {
-    algo_type = EVP_sha256();
+    algo_name = "SHA256";
   }
   else {
-    algo_type = EVP_get_digestbynid(algo_nid);
-    if(!algo_type) {
-      algo_name = OBJ_nid2sn(algo_nid);
-      failf(data, "Could not find digest algorithm %s (NID %d)",
-            algo_name ? algo_name : "(null)", algo_nid);
+    algo_name = OBJ_nid2sn(algo_nid);
+    if(!algo_name) {
+      failf(data, "Could not find digest algorithm name for NID %d",
+            algo_nid);
       return CURLE_SSL_INVALIDCERTSTATUS;
     }
   }
 
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+  algo_type = EVP_MD_fetch(NULL, algo_name, NULL);
```

Comment:
It’s best to cache the result of this, for performance reasons.

Reply:
Hi @DemiMarie! Thank you for continued review.
Where would it be best to cache this? I am honestly not too familiar with this level of curl internals. I am primarily working this as a continuation of the great work done by @steffe-kiess (#7196) to add support for SPNEGO GSSAPI, as it is something we critically need from command-line usage due to some web servers now enforcing this. Any pointer in right direction of caching architecture would be greatly appreciated.

I have found little material online regarding the performance impact of repeatedly calling `EVP_MD_fetch` too much. The explicit fetching will only be done once per connection, during initial connect, so for most cases this function will only be called once. This is a also a completely new feature, as curl has been unable to ever connect to servers requiring secure channel binding. Therefore wondering if caching is something which could be done in a future patch if it turns out to be too much work.

Finally, regarding Explicit Fetching, I have not found any other usage of `EVP_*_fetch` in the entire curl source code... All other places use `EVP_sha256()` and similar implicit functions. I agree it could be a good idea to support explicit fetching, but I'd ideally not want this change to be a testing ground for this migration. Rather reverting back to the implicit-only implementation and then adding OpenSSL 3.x Explicit Fetching support in a separate PR as it touches many more areas outside of this scope. Thoughts?

Reply:
Yup, separate PR.

Reply:
OK! Do you wanna do explicit+caching in a separate PR, or only caching?

Reply:
I’m not a curl developer, just someone who noticed this PR while reading about the mess that OpenSSL 3.x is causing for HAProxy.